File: Data Structures & Algorithms/rotting-fruit/submission-5.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        // Time and Space Complexity: O(row * col)
        int fresh{0};
        int time{0};
        set<pair<int, int>> visited;
        deque<pair<int, int>> q;
        set<pair<int, int>> moves{{0,1},{1,0},{0,-1},{-1,0}};
        for (int i = 0; i< grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                if (grid[i][j] == 1) fresh++;
                else if (grid[i][j] == 2) q.push_back({i,j});
            }
        }
        while (!q.empty() and fresh) {
            int layer = q.size();
            for (int i = 0; i < layer; i++){
                auto [r, c] = q.front();
                q.pop_front();
                for (auto [dr, dc] : moves) {
                    int nr{r + dr}, nc{c + dc};
                    if (min(nr, nc) >= 0 && nr < grid.size() and nc < grid[nr].size() and !visited.contains({nr, nc})) {
                        visited.insert({nr, nc});
                        if (grid[nr][nc] == 1) {
                            grid[nr][nc] = 2;
                            fresh--;
                            q.push_back({nr, nc});
                        }
                    }
                }
            }
            time++;       
        }
        return !fresh ? time : -1;
    }
};

File: Data Structures & Algorithms/the-maze/submission-5.py
class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        # Time Complexity: O(rows * cols)) Space Complexity: O(rows * cols)
        moves = ((0,1),(1,0),(0,-1),(-1,0))
        q = deque()
        visited = set()
        q.append(start)
        while q:
            r, c = q.popleft()
            if [r, c] == destination: return True
            for dr, dc in moves:
                nr, nc = r + dr, c + dc
                while min(nr, nc) >= 0 and nr < len(maze) and nc < len(maze[nr]) and maze[nr][nc] != 1:
                    nr += dr
                    nc += dc
                nr -= dr
                nc -= dc
                if (nr, nc) not in visited:
                    q.append([nr, nc])
                    visited.add((nr, nc))
        return False